### PR TITLE
Mg/validation context 

### PR DESCRIPTION
**Pull request by mangei**

### CODE REVIEW

File: src/main/javascript/service/cat-validation-service.js [hunk 1]
```diff
@@ -15,54 +17,122 @@ angular.module('cat.service.validation', ['cat.service.message'])
...
+            } else {
+                return catValidations;
             }
+        };
+
+        this.createContext = function () {
+            var uuid = window.cat.util.generateUUID();
+            catValidationContexts[uuid] = {
+                uuid: uuid,
+                global: undefined,
```
Comment (tscheinecker) [inline src/main/javascript/service/cat-validation-service.js]: this is dangerous -> see comment below [https://github.com/Catalysts/cat-angular/pull/37/files#diff-917ef52d37b2e2a8b6f242de4aeb8694R117]

File: src/main/javascript/service/cat-validation-service.js [hunk 2]
```diff
@@ -15,54 +17,122 @@ angular.module('cat.service.validation', ['cat.service.message'])
...
             if (!!rejection.data.fieldErrors) {
                 // group by field
                 _.forEach(rejection.data.fieldErrors, function (fieldError) {
-                    fieldErrors[fieldError.field] = fieldErrors[fieldError.field] || [];
-                    fieldErrors[fieldError.field].push(fieldError.message);
+                    // Allow config to switch between displaying errors at the field and displaying errors at known fields or globally
+                    if (catMessagesConfig.knownFieldsActive === true) {
+                        // If the error is for a known field, show the error at the field.
+                        // If not, display it as a global error.
+                        if (_.indexOf(validations.knownFields, fieldError.field) > -1) {
```
Comment (tscheinecker) [inline src/main/javascript/service/cat-validation-service.js]: use `_.contains` or `_.includes` -> https://lodash.com/docs#includes

File: src/main/javascript/service/cat-validation-service.js [hunk 3]
```diff
@@ -15,54 +17,122 @@ angular.module('cat.service.validation', ['cat.service.message'])
...
+            delete this.getValidations(context).global;
+            this.getValidations(context).fieldErrors = {};
         };
 
-        this.clearValidationErrors = function () {
-            delete catValidations.global;
-            catValidations.fieldErrors = {};
+        this.hasGlobalErrors = function (context) {
+            var globalErrors = this.getValidations(context).global;
+            return !!globalErrors && globalErrors.length > 0;
```
Comment (tscheinecker) [inline src/main/javascript/service/cat-validation-service.js]: globalErros might be undefined, which results in an error when trying to evaluate `globalErrors.length`

**Comment by tscheinecker** [comment]
in general i'm not very sure about the necessity for this change

a field error by definition has to be an error associated with an input field, a global error reflects a general error state (of either the application itself or a generic failure in the model, unspecific to a certain field)

could you please give me an example of which error you would consider a field error in the backend, but not have an appropriate input field in your ui?

File: src/main/javascript/service/cat-validation-service.js (hunk 3 above)
Comment (mangei) [inline src/main/javascript/service/cat-validation-service.js]: we check with `!!globalErrors` if the `globalErrors` exists (is not `undefined`). so this does not lead to an error.

**Comment by mangei** [comment]
in general it was necessary to introduce _validation groups_, because we had multiple forms and we had to map the response to the corresponding input element. If you need more details on that, I will give you a detailed answer.

But I think you comment is more related to the _unknown fields_ functionality. I consulted @mdopplinger. He will give an answer on that question.
What I know is, that we had the case, that an entity was extended in the backend by some developer with some constraints, but the input field in the frontend was not added. The frontend did not show any message (because there was no error message assignment for the missing field). If this is the case, with the new functionality, the error is immediately visible by printing the error message of the missing field as a global message.

File: src/main/javascript/service/cat-validation-service.js (hunk 3 above)
Comment (tscheinecker) [inline src/main/javascript/service/cat-validation-service.js]: true, i missed that somehow :)

**Comment by tscheinecker** [comment]
ok, i think i understand what you want to achieve and i'm pretty sure it works out quite fine
one thing though: i would like to see some tests ;)

**Comment by mangei** [comment]
- add tests for validation service
- add documentation
- refactor for better naming
- extract functions/constuctors for better IDE support
- rebased on latest master

**Pull request merged**